### DISCUSSION TITLE
MQTT commands wrong response

### DISCUSSION THREAD

**Discussion by nsitt** [arendst/Tasmota#10561] [Q&A] [answered]
Please can you help me with this issue?

I have a few devices configured and unable to send MQTT commands from one to another.

Here is an example from 2 devices:

Device 1, Display
Device 2, Solar

fulltopic for both is set : %prefix%/%topic%/

Sending mqtt command from Display to Solar is not giving the desired response, instead it is returning a response as if the command was executed localy

14:52:15 CMD: cmnd/PowR2PV/POWER
14:52:15 MQT: stat/Display/RESULT = {"POWER":"OFF"}
14:52:15 MQT: stat/Display/POWER = OFF
14:52:18 CMD: status
14:52:18 MQT: stat/Display/STATUS = {"Status":{"Module":0,"DeviceName":"Dis","FriendlyName":["Display"],"Topic":"Display","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0}}
14:52:21 CMD: status 6
14:52:21 MQT: stat/Display/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.178.5","MqttPort":1883,"MqttClientMask":"Display","MqttClient":"Display","MqttUser":"me","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30}}

Here are the settings of Device 2

14:46:39 CMD: status
14:46:39 MQT: stat/PowR2PV/STATUS = {"Status":{"Module":43,"FriendlyName":["Solar"],"Topic":"PowR2PV","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":0,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0}}
14:46:41 CMD: status 6
14:46:41 MQT: stat/PowR2PV/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.178.5","MqttPort":1883,"MqttClientMask":"PowR2_PV","MqttClient":"PowR2_PV","MqttUser":"me","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30}}

I'd really appreciate your support,

**Comment by barbudor** [comment]
Typing a MQTT topic in the console will not work
You have to use `Publish`

**Reply by nsitt** [reply]
> Typing a MQTT topic in the console will not work
> You have to use `Publish`

From documentation I found the following:
Publish | <topic> <payload> = MQTT publish any topic and optional payload

My attempts to use published failed. What would be the topic and what is the payload? Can you give me an example on how to use it?

16:36:18 CMD: publish display/power
16:36:18 MQT: display/power = 
16:40:13 CMD: publish display cmnd/display/power
16:40:13 MQT: display = cmnd/display/power
16:40:43 CMD: publish cmnd/display/power
16:40:43 MQT: cmnd/display/power = 
16:41:00 CMD: publish display cmnd/display/power
16:41:00 MQT: display = cmnd/display/power
16:41:24 CMD: publish display tele/display/power
16:41:24 MQT: display = tele/display/power
16:41:43 CMD: publish display/cmnd/display/power
16:41:43 MQT: display/cmnd/display/power =

**Reply by barbudor** [reply]
this is the correct syntax : `publish cmnd/display/power` to resquest power state.
To change the power state you need to provide the payload : `publish cmnd/display/power on`  (or off  or toggle or 0 or 1)

**Reply by barbudor** [reply]
But does your display as a Relay ?
Sending a POWER command to a device without a Relay should just make a command error in the Display's console

**Reply by barbudor** [reply]
If you want to send a power command from the display to the PowR2PV, you need to enter `publish cmnd/PowR2PV/power on` on the display

**Reply by meingraham** [reply]
@nsitt 

You may need to spend some time on an MQTT tutorial which will give you the understanding of topics and how they are used in a publish/subscribe manner to communicate with nodes in your IoT network. Subsequently, you need to look at how Tasmota formats topics so that nodes subscribe to the "relevant" topics and are then able to process the information they receive from publishing nodes. Without this basic foundation, you will struggle to make this all work.

**Reply by nsitt** [reply]
> If you want to send a power command from the display to the PowR2PV, you need to enter `publish cmnd/PowR2PV/power on` on the display

Tried all of that, it is not working, I checked on PowR2PV device. Not changing the status

16:56:06 CMD: publish cmnd/PowR2PV/power off
16:56:06 MQT: cmnd/PowR2PV/power = off
16:56:14 CMD: publish cmnd/PowR2PV/power on
16:56:14 MQT: cmnd/PowR2PV/power = on

**Reply by nsitt** [reply]
> But does your display as a Relay ?
> Sending a POWER command to a device without a Relay should just make a command error in the Display's console

Don't get confused by the name.
Yes it has a relay.

16:59:12 CMD: Power 1
16:59:12 MQT: stat/Display/RESULT = {"POWER":"ON"}
16:59:12 MQT: stat/Display/POWER = ON

**Reply by barbudor** [reply]
You need to look at your MQTT broker, may be use a tool like MQTT explorer to understand what is happening
Is it the sender which publication do not work ?
is it the receiver that do not receive the command ?

**Reply by nsitt** [reply]
> @nsitt
> 
> You may need to spend some time on an MQTT tutorial which will give you the understanding of topics and how they are used in a publish/subscribe manner to communicate with nodes in your IoT network. Subsequently, you need to look at how Tasmota formats topics so that nodes subscribe to the "relevant" topics and are then able to process the information they receive from publishing nodes. Without this basic foundation, you will struggle to make this all work.

Yes, thank you for the advise. I spent some time to understand it but I am still strugling. I am looking for some examples on how to configure / setup tasmmota devices so that MQTT commands between them are executed. Unfortunately the documentation and the info in the web is nto a great help. I was hoping that someone can help me who has a running setup and point me in the right direction

**Comment by sfromis** [comment, marked as answer]
Tasmota by default subscribes to the topic `cmnd/devicename/#` meaning that any command like topic `cmnd/devicename/Power` and payload on should work. From another Tasmota device, the command would be `publish cmnd/otherdevice/Power 0`

**Reply by nsitt** [reply]
Yes, that is what I expected from reading the documentation. Unfortunately it is not working in my environment. I am able to control all devices from IoBroker, these are also using MQTT.

**Reply by barbudor** [reply]
So if you can control your devices from IoBroker, your MQTT environment is likely to be correct.
That means that devices can receive proper commands.
Test 1 : Sending a `cmnd/display/power on` from a pure MQTT tool to check if the device can receive the command from another sotware
You should get a tool to send MQTT commands such as MQTT explorer. Or may be IoBroker provides some command line equivalent to "mosquitto_pub" and "mosquitto_sub" that are provided with mosquitto broker.
Test 2 : On a device `publish cmnd/display/power on` and check in the tool if you can see this command passing. That allows to check that your device can actuallly publish to the MQTT
